src/lib.rs

```diff
@@ -195,7 +207,56 @@ fn dag_longest_path_length(
     graph: &digraph::PyDiGraph,
     weight_fn: Option<PyObject>,
 ) -> PyResult<usize> {
-    let (_, path_weight) = longest_path(py, graph, weight_fn)?;
+    let edge_weight_callable =
+        |source: usize, target: usize, weight: &PyObject| -> PyResult<usize> {
+            match &weight_fn {
+                Some(weight_fn) => {
+                    let res = weight_fn.call1(py, (source, target, weight))?;
+                    res.extract(py)
+                }
+                None => Ok(1),
+            }
+        };
+    let (_, path_weight) = longest_path(graph, edge_weight_callable)?;
+    Ok(path_weight)
+}
+
+/// Find the length of the weighted longest path in a DAG
+///
+/// This function differs from :func:`retworkx.dag_longest_path_length` in that
+/// this function requires a ``weight_fn`` parameter, and the ``weight_fn`` is
+/// expected to return a ``float`` not an ``int``.
+///
+/// :param PyDiGraph graph: The graph to find the longest path on. The input
+///     object must be a DAG without a cycle.
+/// :param weight_fn: A python callable that if set will be passed the 3
+///     positional arguments, the source node, the target node, and the edge
+///     weight for each edge as the function traverses the graph. It is expected
+///     to return an float weight for that edge. For example,
+///     ``dag_longest_path(graph, lambda: _, __, weight: weight)`` could be
+///     use to just use an float edge weight. It's also worth noting that this
+///     function traverses in topological order and only checks incoming edges to
+///     each node.
+///
+/// :returns: The longest path length on the DAG
+/// :rtype: float
+///
+/// :raises Exception: If an unexpected error occurs or a path can't be found
+/// :raises DAGHasCycle: If the input PyDiGraph has a cyc
+
+#[pyfunction]
+#[pyo3(text_signature = "(graph, weight_fn, /)")]
+fn dag_weighted_longest_path_length(
+    py: Python,
+    graph: &digraph::PyDiGraph,
+    weight_fn: PyObject,
+) -> PyResult<f64> {
+    let edge_weight_callable =
+        |source: usize, target: usize, weight: &PyObject| -> PyResult<f64> {
+            let res = weight_fn.call1(py, (source, target, weight))?;
+            res.extract(py)
```

Comment:
I probably should add a check here for `NaN` and raise because I don't think the `max_by()` call in `longest_path` will work with it (and I assume panic).

Reply:
Done in: https://github.com/Qiskit/retworkx/pull/393/commits/45f3fb1261df51c1123454790cb8438bc70eaca7